### PR TITLE
Bring SDK Expression types into alignment with CLI + unit tests for Expression

### PR DESCRIPTION
**Pull request by Berlioz**
Writing tests revealed, uh. A number. Of bugs in the current Expression implementation:

0) (big feature we forgot) **a param's default needs to be an Expression in addition to a literal**
1) (feature alignment) comparisons should be able to accept another Expression as the rhs
2) (feature alignment) ternary operators should be able to provide Expressions as the if_true/if_false values
3) (bug) ternary operators should not be forced to be of the same type as the params being compared (i.e `{{ params.INT2 == params.INT2 ? params.STRING : "foo" }}` was illegal)
4) (bug) boolean params would always evaluate to true no matter what string value the CLI passed in
5) (feature alignment) we should support != as a comparison

### CODE REVIEW

File: spec/params/params.spec.ts [hunk 1]
```diff
@@ -0,0 +1,212 @@
...
+    const sameInt = params.defineInt("SAME_INT");
+    const diffInt = params.defineInt("DIFF_INT");
+    expect(int.equals(sameInt).value()).to.be.true;
+    expect(int.equals(-11).value()).to.be.true;
+    expect(int.equals(diffInt).value()).to.be.false;
+    expect(int.equals(22).value()).to.be.false;
+  });
+
+  it("can use all the comparison operators when explicitly requested", () => {
+    const jkl = params.defineString("DIFF_STRING");
```
Comment (taeold) [inline spec/params/params.spec.ts:85]: does comparison operator support comparing params?

e.g. does this work?

```
const asdf = params.defineString("A_STRING")
expect(jkl.cmp(">", asdf).value()).to.be.true;
```

File: spec/params/params.spec.ts [hunk 2]
```diff
@@ -0,0 +1,212 @@
...
+    expect(boolParam.value()).to.be.true;
+
+    const floatParam = params.defineFloat("PI");
+    expect(floatParam.value()).to.equal(3.14159);
+
+    const falseParam = params.defineBoolean("FALSE");
+    expect(falseParam.value()).to.be.false;
+  });
+
+  it("falls back on the javascript zero values in case of type mismatch", () => {
```
Comment (taeold) [inline spec/params/params.spec.ts:58]: hmm fallbacks are tricky. We don't expect these to happen in real life no? (b/c cli will ensure that the values exist)

File: src/params/types.ts [hunk 3]
```diff
@@ -183,26 +214,54 @@ export abstract class Param<T extends string | number | boolean | string[]> exte
...
       name: this.name,
-      ...this.options,
+      ...otherOptions,
       type: (this.constructor as typeof Param).type,
     };
 
-    return out as ParamSpec<T>;
+    if (paramDefault instanceof Expression) {
+      out.default = paramDefault.toCEL();
+    } else if (paramDefault) {
```
Comment (taeold) [inline src/params/types.ts:260]: what is `paramDefault = false`? Should this condition be instead `paramDefault !== undefined` or something?

**Review by taeold — APPROVED** [review]
few comments, mostly good to my eyes.

File: src/params/types.ts (hunk 3 above)
Comment (Berlioz) [reply src/params/types.ts:260]: yeah, that's what I was meaning to capture

File: spec/params/params.spec.ts (hunk 2 above)
Comment (Berlioz) [reply spec/params/params.spec.ts:58]: Correct.

File: spec/params/params.spec.ts (hunk 1 above)
Comment (Berlioz) [reply spec/params/params.spec.ts:85]: It does (that was the impact of changing the `rhs` type in a ComparisonExpression to be `rhs: T | Expression<T>`) but you're right in that I should add tests

**Pull request merged**